test/unit/Monitoring/StatsiteMetricsAgentTest.php

```diff
@@ -0,0 +1,100 @@
+{
+    private array $messages = [];
+
+    public function getMessages(): array
+    {
+        return $this->messages;
+    }
+
+    protected function sendToSocket(string $message): void
+    {
+        $this->messages[] = $message;
+    }
+}
+
+class StatsiteMetricsAgentTest extends TestCase
+{
+    public function test_it_is_initialisable()
+    {
+        $this->assertInstanceOf(StatsiteMetricsAgent::class, $this->newSubject());
+    }
+
+    public function test_it_increments_counter_by_one()
+    {
+        $subject = $this->newSubject();
+        $subject->incrementCounterByOne(MetricId::nameAndSource('logins', 'test'));
+        $subject->incrementCounterByOne(MetricId::nameAndSource('login_failed', 'test'));
+        $subject->incrementCounterByOne(MetricId::nameAndSource('logins', 'test'));
+        $this->assertSame(['logins=test:1|c', 'login_failed=test:1|c', 'logins=test:1|c'], $subject->getMessages());
+    }
+
+    public function test_it_sets_gauge()
+    {
+        $subject = $this->newSubject();
+        $subject->setGauge(MetricId::nameAndSource('active_sessions', 'test'), 205);
+        $subject->setGauge(MetricId::nameAndSource('active_sessions', 'test'), 108);
+        $this->assertSame(['active_sessions=test:205|g', 'active_sessions=test:108|g'], $subject->getMessages());
+    }
+
+    public function test_it_records_counter_integral()
+    {
+        $subject = $this->newSubject();
+        $subject->recordCounterIntegral(MetricId::nameAndSource('opcache_oom', 'test'), 2);
+        $this->assertSame(['opcache_oom=test:2|kv'], $subject->getMessages());
+    }
+
+    public function test_it_adds_timer()
+    {
+        $subject = $this->newSubject();
+        $subject->addTimer(
+            MetricId::nameAndSource('email_sent', 'test'),
+            new DateTimeImmutable('2020-03-30 10:01:02.3455'),
+            new DateTimeImmutable('2020-03-30 10:01:03.5000')
+        );
+        $this->assertSame(['email_sent=test:1154.5|ms'], $subject->getMessages());
+    }
+
+    public function test_it_adds_sample()
+    {
+        $subject = $this->newSubject();
+        $subject->addSample(MetricId::nameAndSource('mem_used', 'test'), 17.24);
+        $this->assertSame(['mem_used=test:17.24|ms'], $subject->getMessages());
+    }
+
+    public function test_it_replaces_metric_source_when_required()
+    {
+        $subject = $this->newSubject();
+        $subject->setSourceHostname('app-server-321');
+
+        $subject->incrementCounterByOne(MetricId::forHost('my_metric'));
+        $subject->incrementCounterByOne(MetricId::nameOnly('lonely_metric'));
+        $subject->incrementCounterByOne(MetricId::nameAndSource('other', 'system'));
+        $this->assertSame(
+            [
+                'my_metric=app-server-321:1|c',
+                'lonely_metric=:1|c',
+                'other=system:1|c',
+            ],
+            $subject->getMessages()
+        );
+    }
+
```

Comment:
Just a thought, given we do a bit of parsing on the incoming `gethostname()` by default, ideally we'd test that... Maybe instead of / as well as above, given you already have the `SpyingStatsiteMetricsAgent` it'd be worth pulling the actual `gethostname()` call to a protected func that could be stubbed in the spy, same as you've done for `socket_sendto`?

Reply:
Yeh, also need to find out if it's true that statsite doesn't support periods in the source name and probably update the setter to replace them with hyphens.

Reply:
not be able to just wrap it in a method of our own as it's called from the constructor so I can't replace it before I construct it. It'll have to become an optional constructor arg I think?

Reply:
Only thing then is does the optional constructor arg get parsed same way we parse hostname, or does it skip parsing and store exactly what you pass the same way that your `setHostname` does at the moment?

I think you could still wrap it if you wanted, albeit with a slightly nasty static prop, something like:

```
class StatsiteAgent {
  __construct() {
    $host = $this->getHostname();
    $this->host = explode('.', $host); // or whatever
  }

  protected function getHostname() {
    return gethostname();
  }
}

class SpyingAgent {
  private static $stub_hostname = 'fake';
  public static function withStubbedHostname($host) {
    static::$stub_hostname = $host;
    return new static;
  }
  
  protected function getHostname() {
    return static::$stub_hostname;
  }
}
```

Reply:
well given the class has a setter for setting hostname, I think the parsing should also apply to the setter and we could use the setter in the constructor?

Reply:
I guess the thought behind being able to set it was so if you did want to override it but not from within your dependency container then you could do it elsewhere without having to factory it through another class, but don't know which would be better

Reply:
yeah, parse in the setter too and just use / test that.